[lldb] Store the command in the CommandReturnObject

JDevlieghere opened the pull request:
As suggested in #125006. Depending on which PR lands first, I'll update `TestCommandInterepterPrintCallback.py` to check that the `CommandReturnObject` passed to the callback has the correct command.

jimingham commented:
This looks good.  I wonder if we ought to do something a little less ad hoc about the cases where our command return objects are from HandleCommands, so they won't necessarily have the result of a single command.  In the case where a UI is parsing these, it might very well need to know that fact.  They all currently have `<>` around the command name, but sadly we don't outlaw `<` as the initial character of a command, so you can't 100% tell from that.

JDevlieghere commented:
> This looks good. I wonder if we ought to do something a little less ad hoc about the cases where our command return objects are from HandleCommands, so they won't necessarily have the result of a single command. In the case where a UI is parsing these, it might very well need to know that fact. They all currently have `<>` around the command name, but sadly we don't outlaw `<` as the initial character of a command, so you can't 100% tell from that.

I can add an enum to indicate whether the command was a user command or something LLDB put together and have two constructors that each do the appropriate thing. `GetCommand` could then return a predefined value for non-user commands.

labath reviewed (commented):
This idea has bothered me from the beginning, but I didn't want to say anything, as it was mostly just a feeling. After seeing this patch, I think I can put some words behind it.

The thing I don't like about this patch is the repetition/redundancy it creates. Like, you first create the CommandReturnObject, give it the command it's hold the result of. Then, you pass that object into HandleCommand, but that's not all you give it -- the function takes the string of the command, again.

It's not the end of the world, but I think it's unfortunate, and while it works nice for your current use case, I don't think it makes that much sense for other usages. For example, if the `CommandReturnObject` had always contained the command part, would the signature of a python command still be `def cmd(debugger, command, result, dict):`, or would we have expected the command to retrieve it from the result object?

It's hard to say, but what I think we can say is that if you make your current callback take the "command" and the "result of that command" as two separate arguments, then it will at least be consistent with the signature above.

jimingham commented:
One way to make the intent of the command string in the return object clear is to ONLY set it after we've called HandleCommand (or HandleCommands).  That way it's explicitly NOT meant to help execute the command - after all, commands really should NOT be operating on the un-alias-resolved forms of their commands.  Then the intent will be clear.  We could even call the API `GetOriginalCommandString` to make it doubly clear.

JDevlieghere committed:
[lldb] Store the command in the CommandReturnObject

As suggested in #125006. Depending on which PR lands first, I'll update
TestCommandInterepterPrintCallback.py to check that the
CommandReturnObject passed to the callback has the correct command.

JDevlieghere commented:
Rebased and implemented @jimingham's latest suggestion.

jimingham commented on lldb/include/lldb/API/SBCommandReturnObject.h:
```diff
@@ -42,6 +42,8 @@ class LLDB_API SBCommandReturnObject {
 
   bool IsValid() const;
 
+  const char *GetCommand();
```
The docs currently in CommandReturnObject.h should certainly be here (or in both places)...

JDevlieghere committed:
Add docstring

jimingham commented:
Other that the comment comment, LGTM.

jimingham reviewed (approved):
LGTM

labath commented:
I think this is better. As much as I like initializing things in the constructor, I don't think it makes sense when the whole CommandReturnObject is built around using it as a by-ref return value.

Michael137 commented:
Btw, there are now two tests in the LLDB test-suite called `TestSBCommandReturnObject.py`:
```
./lldb/test/API/api/command-return-object/TestSBCommandReturnObject.py
./lldb/test/API/python_api/commandreturnobject/TestSBCommandReturnObject.py
```
which `lldb-dotest` complains about:
```
  File "/home/llvm-project/lldb/packages/Python/lldbsuite/test/dotest.py", line 704, in visit
    raise Exception("Found multiple tests with the name %s" % name)
Exception: Found multiple tests with the name TestSBCommandReturnObject.py
```

Could you rename one of them please?

JDevlieghere commented:
Merged them in 472220077383b2dbd9cfcaffcc6030558ba7a744
